File: python/graphs/1765-map-of-highest-peak.py
from collections import deque
from typing import List


def highestPeak(isWater: List[List[int]]) -> List[List[int]]:
    height = [[0 for _ in range(len(isWater[0]))] for _ in range(len(isWater))]
    ways_x, ways_y = [-1, 1, 0, 0], [0, 0, 1, -1]
    
    queue = deque()
    
    for i in range(len(isWater)):
        for j in range(len(isWater[0])):
            if isWater[i][j] == 1:
                queue.append((i, j, 0))
                
    while queue:
        x, y, h = queue.popleft()

        for delta_x, delta_y in zip(ways_x, ways_y):
            new_x, new_y = x + delta_x, y + delta_y
            if 0 <= new_x < len(isWater) and 0 <= new_y < len(isWater[0]) and height[new_x][new_y] == isWater[new_x][new_y] == 0:
                height[new_x][new_y] = h + 1
                queue.append((new_x, new_y, h + 1))
                            
    return height

File: cpp/graphs/1765-map-of-highest-peak.cpp
#include <queue>
#include <tuple>
#include <vector>

using namespace std;


vector<vector<int>> highestPeak(vector<vector<int>>& isWater) {
    vector<vector<int>> height(isWater.size(), vector<int> (isWater[0].size()));
    
    queue<tuple<int, int, int>> q;
    vector<vector<int>> ways = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    
    for (int i = 0; i < isWater.size(); i++) {
        for (int j = 0; j < isWater[0].size(); j++) {
            if (isWater[i][j] == 1) q.push({i, j, 0});
        }
    }
    
    while (!q.empty()) {
        tuple<int, int, int> cur = q.front();
        q.pop();
        
        
        for (const auto way : ways) {
            int new_x = get<0>(cur) + way[0];
            int new_y = get<1>(cur) + way[1];

            if (new_x >= 0 && new_x < isWater.size() && new_y >= 0 && new_y < isWater[0].size() &&
                height[new_x][new_y] == 0 && isWater[new_x][new_y] == 0) {
                height[new_x][new_y] = get<2>(cur) + 1;
                q.push({new_x, new_y, get<2>(cur) + 1});
            }
        }
    }
    
    return height;
}